### ISSUE TITLE
Migrating to MMCV version 2.x & Nvidia Blackwell B200 GPUs

### ISSUE THREAD

**Issue by deepak-vij** [OpenDriveLab/UniAD#263]
Hi @hli2020 @eltociear @faikit @Yihanhu @wljungbergh , after jumping through hoops, I finally got the the entire UniAD2.0 training/eval environment (BevFormer, Stage1, Stage2) working on Nvidia Blackwell B200 environment. However, I ended up using MVCC libraries version 1.6.1 compatible libraries. I tried switching to MVCC 2.x codebase. However, due to major code refactoring, I gave up and ended up staying with version 1.6.1.

The question I have is, has anyone in the community tried the migration effort from MVCC version 1.x towards version 2.x. It looks like substantial amount of migration work due to major code refactoring. Although, benefits of doing so seem very tempting (optimized cuda kernels, support for tocrh.compile, kernel fusion etc.).

Please let me know your thoughts before I start thinking through this migration effort. Appreciate it. Thanks.

**Comment by niouoklul** [comment]
Hi @deepak-vij and UniAD maintainers ,

Thanks for starting this discussion about migrating UniAD 2.0 to MMCV 2.x on GPUs like the Blackwell B200 — I’m encountering a similar situation and would love some detailed guidance.

Specifically, would it be possible for you or anyone in the community to provide:

1. The **exact set of installation steps** you used to get UniAD training / evaluation working on the Blackwell B200 (including:
   - PyTorch + CUDA versions
   - exact mmcv/mmcv‑full version and how it was installed
   - mmdet, mmdet3d, mmseg versions that worked
   - any required patch files / configuration changes
   - whether openmim was used or manual wheel installs

2. If you attempted migration to MMCV 2.x, could you share:
   - where the migration breaks (specific files / APIs)
   - which parts of UniAD code depend on MMCV 1.x internals
   - pointers on how to update those parts to MMCV 2.x (example diffs or a patch)

3. Any **working environment configuration** (environment.yaml or requirements list) that others can reproduce easily.

I’m open to testing combinations and providing logs if that helps — any concrete steps or best practices would be much appreciated!

Thanks!

**Comment by deepak-vij** [comment]
@niouoklul and rest of the community. Sorry about not responding earlier, have been awefully busy these days. In any case, in order to make UniAD work with Nvidia-Blackwell B200, I had to jump through lots of hoops to make it all work. Here it goes, hope it helps.
++++++++++++++++
**Dependencies**: The UniAD codebase lists specific versions of mmcv, mmdet, mmdet3d, and mmsegmentation as required dependencies for installation and operation, emphasizing its reliance on specific versions of the MM-series libraries. I encountered major version compatibility related issues while setting up UniAD.

**MMCV 2.x**:
- Based on the installation instructions, pip install instruction for installing mmcv library (version 1.6.1) depends on "torch cuda" version 11.8. 
- As "torch cuda" version 11.8 is incompatible with Blackwell B200 GPU hardware environment, I decided to build mmcv library (version 1.6.1) from the source code as a result. FORCE_CUDA=1 MMCV_WITH_OPS=1 MMCV_CUDA_ARGS="$MMCV_CUDA_ARGS" TORCH_CUDA_ARCH_LIST="$TORCH_CUDA_ARCH_LIST" pip install -v -e . --no-build-isolation 2>&1 | tee /tmp/mmcv_build.log
- Build process from the source code for mmcv library (version 1.6.1) kept failing. Decided to use the latest newer version of mmcv library (version 2.x) instead. The build process for mmcv library (version 2.x) was successful. Afterwards, the compatible versions of other dependent libraries mmdet (version 3.2.0), mmseg (version 1.2.2), mmdet3d (version 1.4.0), mmengine (version 0.10.7) were installed successfully. This was to ensure version compatibility across all of these libraries.
- However, the UniAD2.0 evaluation example script, that verifies the correctness of the prepared environment, kept failing due to UniAD codebase incompatibility with newer mvcc library (version 2.x). It turned out the API structure for the newer mvcc library (version 2.x) has been entirely overhauled. In order to make it all work would involve major code migration effort in the UniAD2.0 codebase. This is not desirable at this moment.

**MMCV 1.x:**
- So the only option left was to strictly use the versions as described in the UniAD2.0 installation instructions. However, the dilemma, as mentioned earlier, was that I was unable to build the older version of the mvcc library (version 1.6.1), it kept failing for some reason.
- It turned out that the build process of the mvcc library (version 1.6.1) from the source code was failing due to  C++ standards related issues. Building MMCV 1.6.1 with CUDA 12.8 is a significant technical challenge because MMCV 1.x was written for the C++ standards and CUDA APIs of the CUDA 10/11 era. CUDA 12 introduces "breaking" changes, particularly in how it handles headers and C++ standards.
- The solution for resolving this issue was to change the setup.py file (sed -i 's/-std=c++14/-std=c++17/g' setup.py). This command finds a specific string in the Python setup file and replaces it with another. In this context, it is being used to upgrade the C++ standard used to compile a Python extension (like UniAD's custom CUDA operators). With this change, the build process for mmcv library (version 1.6.1) worked perfectly.
- pip install mmdet3d==1.0.0rc6 needs to be done using the "--no-build-isolation" option. pip install mmdet3d==1.0.0rc6 --no-build-isolation.
- Additionally, minor code changes were applied to resolve the PyTorch 2.x compatibility related issue.
- **UniAD/tools/test.py** - Fix related to the MMCV / PyTorch 2.x compatibility bug.
import torch.nn as nn
if not hasattr(nn.parallel.DistributedDataParallel, '_use_replicated_tensor_module'):
    &emsp; nn.parallel.DistributedDataParallel._use_replicated_tensor_module = False
- **UniAD/tools/train.py** - Fix related to the MMCV / PyTorch 2.x compatibility bug.
import torch.nn as nn
if not hasattr(nn.parallel.DistributedDataParallel, '_use_replicated_tensor_module'):
    &emsp; nn.parallel.DistributedDataParallel._use_replicated_tensor_module = False

**Comment by deepak-vij** [comment]
Another issue I ran into was related to MMCV 1.6.1 + PyTorch 2.4/2.5 compatibility issue. It happens because in recent versions of PyTorch, the internal function _get_stream expects a torch.device object, but older MMCV (like 1.6.1) passes a raw integer (the GPU ID). In older PyTorch versions, this was fine, but now it fails because an integer doesn't have a .type attribute.

**/mmcv/mmcv/parallel/_functions.py**
import torch
streams = [_get_stream(torch.device(f'cuda:{device}')) for device in target_gpus]

I was getting an Assertion error. This AssertionError happens because of a change in how evaluation results are handled between different versions of MMEngine and MMDetection3D. In older versions of the OpenMMLab suite (like those compatible with MMCV 1.6.1), the evaluation loop expected results to be a simple list of predictions. However, in newer versions or specific configurations, the output is often a dictionary containing different metrics or a list of dictionaries.
The code at line 443 of nuscenes_dataset.py is strictly checking: assert isinstance(results, list), 'results must be a list'. For MMCV library (version 1.6.1), the most direct fix is to modify the nuscenes_dataset.py file to handle the case where results is a dictionary.
**/opt/conda/envs/uniad2.0/lib/python3.9/site-packages/mmdet3d/datasets/nuscenes_dataset.py**
if isinstance(results, dict):
            &emsp; \# Check for common keys
            &emsp; if 'pts_bbox' in results:
                &emsp; &emsp; results = results['pts_bbox']
            &emsp; elif 'img_bbox' in results:
                &emsp;&emsp; results = results['img_bbox']
            \# --- ADD THESE COMMON BEVFORMER KEYS ---
            &emsp; elif 'bbox_results' in results:
                &emsp;&emsp; results = results['bbox_results']
            &emsp; elif 'results' in results and isinstance(results['results'], list):
                &emsp;&emsp; results = results['results']

**Comment by deepak-vij** [comment]
**Use Mini Dataset instead of default Large Trainval Dataset** 
The Trainval is quite a large dataset (1/2 a tera byte). The training process for it would run for couple of days to complete. Instead, I decided to use the subset (mini) of the Trainval dataset. By default, UniAD codebase looks for the Trainval dataset. I used the following python code for UniAD to use subset (mini) dataset instead.

python tools/create_data.py nuscenes \
    --root-path ./data/nuscenes \
    --out-dir ./data/nuscenes \
    --extra-tag nuscenes \
    --version v1.0-mini \
    --canbus ./data/nuscenes

cp "nuscenes_infos_temporal_val.pkl" and "nuscenes_infos_temporal_train.pkl" to data/infos folder

**Comment by deepak-vij** [comment]
**Bevformer Backbone Training & Evaluation Script Modifications** 
\#**Training**
./tools/uniad_dist_train.sh ./projects/configs/bevformer/base_bevformer.py 8
\#**Evaluation**
./tools/uniad_dist_eval.sh ./projects/configs/bevformer/base_bevformer.py /root/UniAD/projects/work_dirs/bevformer/base_bevformer/latest.pth 8

**Stage-1 Training & Evaluation Script Modifications**
\#**Training**
./tools/uniad_dist_train.sh ./projects/configs/stage1_track_map/base_track_map.py 8
\#**Evaluation**
./tools/uniad_dist_eval.sh ./projects/configs/stage1_track_map/base_track_map.py /root/UniAD/projects/work_dirs/stage1_track_map/base_track_map/latest.pth 8 

**Stage-2 Training & Evaluation Script Modifications**
\#**Training**
./tools/uniad_dist_train.sh ./projects/configs/stage2_e2e/base_e2e.py 8
\#**Evaluation**
./tools/uniad_dist_eval.sh ./projects/configs/stage2_e2e/base_e2e.py /root/UniAD/projects/work_dirs/stage2_e2e/base_e2e/latest.pth 8

**Comment by deepak-vij** [comment]
**Baseline Environment Setup & Verification**
1. Started out with docker container using image "nvidia/cuda:12.8.0-devel-ubuntu22.04"
2. Install basic tooling
apt-get update && apt-get install -y \
  git curl wget vim \
  build-essential ninja-build cmake pkg-config \
  libglib2.0-0 libsm6 libxext6 libxrender1 libgl1 \
  && rm -rf /var/lib/apt/lists/*
 
wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O /tmp/miniconda.sh
bash /tmp/miniconda.sh -b -p /opt/conda
rm -f /tmp/miniconda.sh
export PATH=/opt/conda/bin:$PATH

**Install & Verify Torch/Cuda setup**
\# install cu128 wheels
python -m pip install --upgrade pip 
python -m pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu128

\# Verification
python - <<'PY'
import torch
print("torch:", torch.__version__)
print("cuda :", torch.version.cuda)
print("arch list:", torch.cuda.get_arch_list())
print("device:", torch.cuda.get_device_name(0))
print("capability:", torch.cuda.get_device_capability(0))
x = torch.randn(4, 4, device="cuda")
y = x @ x.T
print("matmul ok:", y[0,0].item())
PY

**Verify mmcv libraries setup**
\# This command is a health check for your MMCV installation.
\# It is specifically designed to verify that the C++ and CUDA extensions—which are
\# required for heavy-duty tasks like NMS (Non-Maximum Suppression) in UniAD—were
\# compiled and installed correctly.
python - <<'PY'
import mmcv, glob, site
sp = site.getsitepackages()[0]
print("mmcv:", mmcv.__version__, mmcv.__file__)
print("mmcv _ext:", glob.glob(sp + "/mmcv/_ext*.so"))
print("mmcv ops ext:", glob.glob(sp + "/mmcv/ops/_ext*.so"))
from mmcv.ops import nms
print("mmcv.ops.nms import OK")
PY

\# This command is a comprehensive environment validation script.
\# While your previous script focused on the foundation (MMCV), this one checks
\# the entire OpenMMLab 2.0 stack and tests whether the specialized CUDA operators
\# are accessible across the different libraries. It is particularly useful for 
\# UniAD because UniAD is a "full-stack" perception model that uses components from
\# 2D detection, 3D detection, and segmentation etc.

python - <<'PY'
import mmcv, mmengine, mmdet, mmseg, mmdet3d
print("mmcv:", mmcv.__version__)
print("mmengine:", mmengine.__version__)
print("mmdet:", mmdet.__version__)
print("mmseg:", mmseg.__version__)
print("mmdet3d:", mmdet3d.__version__)
 
\# Force an import path that exercises ops
from mmdet.structures.mask.structures import BitmapMasks
from mmcv.ops import roi_align
print("imports + ops OK")
PY
mmcv: 2.1.0
mmengine: 0.10.7
mmdet: 3.2.0
mmseg: 1.2.2
mmdet3d: 1.4.0
imports + ops OK
